File: app/src/main/java/com/inappropirates/lightwalker/config/ModeManager.java
package com.inappropirates.lightwalker.config;

import com.inappropirates.lightwalker.R;
import com.inappropirates.lightwalker.SettingsActivity;

import java.util.ArrayList;
import java.util.List;

public enum ModeManager
{
    INSTANCE;

    private Mode currentMode = null;
    private List<Mode> modes = new ArrayList<Mode>()
    {
        {
            add(new Mode("main", SettingsActivity.class, R.layout.main_preferences, false));
            add(new MandyMode("mandy", null, null, true));
            add(new Mode("sparkle", SettingsActivity.class, R.layout.sparkle_preferences, true));
            add(new Mode("rainbow", SettingsActivity.class, R.layout.rainbow_preferences, true));
            add(new Mode("zebra", SettingsActivity.class, R.layout.zebra_preferences, true));
            add(new Mode("chaos", SettingsActivity.class, R.layout.chaos_preferences, true));
            add(new Mode("flames", SettingsActivity.class, R.layout.flames_preferences, true));
            add(new Mode("equalizer", SettingsActivity.class, R.layout.equalizer_preferences, true));
            add(new Mode("bubble", SettingsActivity.class, R.layout.bubble_preferences, true));
            add(new Mode("gravity", SettingsActivity.class, R.layout.gravity_preferences, true));
        }
    };

    public Mode getMode(String modeName)
    {
        for (Mode mode : modes)
            if (mode.getName().equals(modeName))
                return mode;

        return null;
    }

    public Mode getCurrentMode() {
        return currentMode;
    }

    public void setCurrentMode(Mode currentMode) {
        this.currentMode = currentMode;
    }

    public List<Mode> getModes() {
        return modes;
    }
}

File: app/src/main/kotlin/com/inappropirates/lightwalker/remote/modes/ModeManager.kt
package com.inappropirates.lightwalker.remote.modes

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.util.Log
import androidx.compose.runtime.mutableStateOf
import androidx.preference.PreferenceManager
import com.inappropirates.lightwalker.remote.ui.MandyActivity
import com.inappropirates.lightwalker.remote.bluetooth.Bt
import com.inappropirates.lightwalker.remote.ui.BubbleConfigActivity
import com.inappropirates.lightwalker.remote.ui.ChaosConfigActivity
import com.inappropirates.lightwalker.remote.ui.EqualizerConfigActivity
import com.inappropirates.lightwalker.remote.ui.FlamesConfigActivity
import com.inappropirates.lightwalker.remote.ui.RainbowConfigActivity
import com.inappropirates.lightwalker.remote.ui.SparkleConfigActivity
import com.inappropirates.lightwalker.remote.ui.ZebraConfigActivity
import com.inappropirates.lightwalker.remote.util.PropertyFormatter

object ModeManager {
    private val TAG = this::class.java.simpleName
    val modes = listOf(
        Mode("main", enabled = false),
        Mode("mandy"),
        Mode("sparkle", configActivity = SparkleConfigActivity::class.java),
        Mode("rainbow", configActivity = RainbowConfigActivity::class.java),
        Mode("zebra", configActivity = ZebraConfigActivity::class.java),
        Mode("chaos", configActivity = ChaosConfigActivity::class.java),
        Mode("flames", configActivity = FlamesConfigActivity::class.java),
        Mode("equalizer", configActivity = EqualizerConfigActivity::class.java),
        Mode("bubble", configActivity = BubbleConfigActivity::class.java),
        Mode("gravity", enabled = false)
    )
    val modeMap: Map<String, Mode> = modes.associateBy { it.name }

    val modeState = mutableStateOf(modeMap["main"]!!)

    suspend fun setMode(context: Context, mode: String) {
        modeMap[mode]?.let {
            setMode(context, it)
        } ?: throw RuntimeException("Mode ${mode} not found!")
    }

    suspend fun setMode(context: Context, mode: Mode) {
        when (mode) {
            modeMap["mandy"] -> {
                Intent(context, MandyActivity::class.java)
                    .also { context.startActivity(it) }
            }

            else -> {
                modeState.value = mode
                sendSettings(context, mode)
            }
        }
    }

    suspend fun sendSettings(context: Context, mode: Mode) {
        if (Bt.connection == null) {
            Log.d(TAG, "sendSettings: no connection available")
            return
        }

        val preferences: SharedPreferences =
            PreferenceManager.getDefaultSharedPreferences(context)

        preferences
            .all
            .entries
            .filter { it.key.startsWith("main") || it.key.startsWith(mode.name) }
            .forEach { pref ->
                pref.value?.let { v ->
                    PropertyFormatter
                        .getStringVal(pref.key, v)
                        .let { Bt.send(pref.key, it) }
                }
            }

        Bt.send("mode", Modes.valueOf(mode.name).ordinal.toString())
    }
}